src/Composer/Package/Archiver/ArchiveManager.php

```diff
@@ -71,36 +71,61 @@ public function setOverwriteFiles(bool $overwriteFiles): self
     }
 
     /**
-     * Generate a distinct filename for a particular version of a package.
-     *
-     * @param CompletePackageInterface $package The package to get a name for
-     *
-     * @return string A filename without an extension
+     * @return array<string, string>
      */
-    public function getPackageFilename(CompletePackageInterface $package): string
+    public function getPackageFilenameParts(CompletePackageInterface $package): array
     {
-        if ($package->getArchiveName()) {
-            $baseName = $package->getArchiveName();
-        } else {
+        $baseName = $package->getArchiveName();
+        if (null === $baseName) {
             $baseName = Preg::replace('#[^a-z0-9-_]#i', '-', $package->getName());
         }
-        $nameParts = [$baseName];
 
-        if (null !== $package->getDistReference() && Preg::isMatch('{^[a-f0-9]{40}$}', $package->getDistReference())) {
-            array_push($nameParts, $package->getDistReference(), $package->getDistType());
+        $parts = [
+            'base' => $baseName,
+        ];
+
+        $distReference = $package->getDistReference();
+        if (null !== $distReference && Preg::isMatch('{^[a-f0-9]{40}$}', $distReference)) {
+            $parts['dist_reference'] = $distReference;
+            $parts['dist_type'] = $package->getDistType();
         } else {
-            array_push($nameParts, $package->getPrettyVersion(), $package->getDistReference());
+            $parts['version'] = $package->getPrettyVersion();
+            $parts['dist_reference'] = $distReference;
+        }
+
+        $sourceReference = $package->getSourceReference();
+        if (null !== $sourceReference) {
+            $parts['source_reference'] = substr(sha1($sourceReference), 0, 6);
         }
 
-        if ($package->getSourceReference()) {
-            $nameParts[] = substr(sha1($package->getSourceReference()), 0, 6);
+        $parts = array_filter($parts);
+        foreach ($parts as $key => $part) {
+            $parts[$key] = str_replace('/', '-', $part);
         }
 
-        $name = implode('-', array_filter($nameParts, static function ($p): bool {
-            return !empty($p);
-        }));
+        return $parts;
+    }
+
+    /**
+     * @param array<string, string> $parts
+     *
+     * @return string
```

Comment:
Let's avoid expanding the public method surface here

```suggestion
     * @return string
     * @internal
```